File: cs/ch17/Money.Tests/BankTest.cs
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using NUnit.Framework;
using TheSoftwareGorilla.TDD.Money;

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion  

[TestFixture]
[Description("BankTest class")]
public class BankTest
{
    private static Bank _bank;

    [OneTimeSetUp]
    public static void OneTimeSetUp()
    {
        _bank = GetBankWithRates();
    }

    public static Bank GetBankWithRates()
    {
        if (_bank == null)
        {
            _bank = new Bank();
            _bank.AddRate("CHF", "USD", 2);
            _bank.AddRate("ZAR", "USD", 17);
            _bank.AddRate("ZAR", "CHF", 20);
        }
        return _bank;
    }

    [TestCase("USD", "USD", 1, TestName = "Rate USD to USD returns 1")]
    [TestCase("CHF", "CHF", 1, TestName = "Rate CHF to CHF returns 1")]
    [TestCase("ZAR", "ZAR", 1, TestName = "Rate ZAR to ZAR returns 1")]
    [TestCase("USD", "CHF", 0, TestName = "Rate CHF to USD missing")]
    [TestCase("USD", "ZAR", 0, TestName = "Rate USD to ZAR missing")]
    [TestCase("CHF", "ZAR", 0, TestName = "Rate CHF to ZAR missing")]   
    [TestCase("CHF", "USD", 2, TestName = "Rate USD to CHF returns rate (2)")]
    [TestCase("ZAR", "USD", 17, TestName = "Rate USD to ZAR returns rate (17)")]
    [TestCase("ZAR", "CHF", 20, TestName = "Rate CHF to ZAR returns rate (20)")]
    [Category("rates")]
    public void TestRate(string from, string to, int expected)
    {
        Assert.That(_bank.Rate(from, to), Is.EqualTo(expected));
    }

    [TestCase("CHF", 2, "USD", 1, TestName = "Reduce CHF 2 to USD 1")]
    [TestCase("ZAR", 20, "CHF", 1, TestName = "Reduce ZAR 20 to CHF 1")]
    [TestCase("ZAR", 17, "USD", 1, TestName = "Reduce ZAR 17 to USD 1")]
    [TestCase("USD", 5, "USD", 5, TestName = "Reduce USD 5 to USD 5")]
    [Category("rates")]
    public void TestReduce(string from, int fromAmt, string to, int expected)
    {
        Money result = _bank.Reduce(MoneyTest.GetCurrencyFactory(from).Invoke(fromAmt), to);
        Assert.That(result, Is.EqualTo(MoneyTest.GetCurrencyFactory(to).Invoke(expected)));
    }

    [TestCase("USD", 1, "ZAR", 17, TestName = "Reduce USD 1 to ZAR 17 throws invalid operation exception")]
    [TestCase("CHF", 1, "ZAR", 20, TestName = "Reduce CHF 1 to ZAR 20 throws invalid operation exception")]
    [TestCase("USD", 1, "CHF", 2, TestName = "Reduce USD 1 to CHF 2 throws invalid operation exception")]
    [Category("rates")]
    public void TestReduceDivisionByZero(string from, int fromAmt, string to, int toRate)
    {
        Exception ex = Assert.Throws<InvalidOperationException>(() => _bank.Reduce(MoneyTest.GetCurrencyFactory(from).Invoke(fromAmt), to));
        Assert.That(ex.Message, Is.EqualTo($"No rate found for {from} to {to}"));
    }

    [TestCase("GBP", 1, "ZAR", 20, TestName = "Reduce GBP 1 to ZAR throws key not found exception")]
    [TestCase("INR", 1, "ZAR", 20, TestName = "Reduce INR 1 to CHF throws key not found exception")]
    [TestCase("FRF", 1, "CHF", 0, TestName = "Reduce FRF 1 to CHF throws key not found exception")]
    [Category("rates")]
    public void TestReduceNotInRateTable(string from, int fromAmt, string to, int toRate)
    {
        Exception ex = Assert.Throws<KeyNotFoundException>(() => _bank.Reduce(MoneyTest.GetCurrencyFactory(from).Invoke(fromAmt), to));
        Assert.That(ex.Message, Is.EqualTo($"The given key '{from}' was not present in the dictionary."));
    }

}

File: java/ch17-03-decimal/src/test/java/com/thesoftwaregorilla/tdd/money/BankTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

//<editor-fold desc="TO DO List">
//</editor-fold>


@DisplayName("Bank Tests")
public class BankTest {

    private static Bank bank;

    @BeforeAll
    public static void beforeAllSetUp() {
        bank = getBankWithRates();
    }

    public static Bank getBankWithRates() {
        if (bank == null) {
            bank = new Bank();
            bank.addRate("CHF", "USD", BigDecimal.valueOf(2)); // 1 CHF = 0.5 USD
            bank.addRate("ZAR", "USD", BigDecimal.valueOf(17)); // 1 ZAR = 0.0588 USD
            bank.addRate("ZAR", "CHF", BigDecimal.valueOf(20)); // 1 ZAR = 0.05 CHF
            bank.addRate("USD", "CHF", BigDecimal.valueOf(0.5)); // 1 CHF = 0.5 USD
            bank.addRate("USD", "ZAR", BigDecimal.valueOf(0.0588235)); // 1 ZAR = 0.0588 USD
        }
        return bank;
    }

    @DisplayName("rate table")
    @ParameterizedTest(name = "rate from = \"{0}\" to = \"{1}\", expected = {2}")
    @CsvSource({
            "USD, USD, 1.00000000",
            "CHF, CHF, 1.00000000",
            "ZAR, ZAR, 1.00000000",
            "USD, CHF, 0.50000000",
            "CHF, ZAR, 0E-8",
            "GBP, ZAR, 0E-8",
            "CHF, USD, 2.00000000",
            "ZAR, USD, 17.00000000",
            "USD, ZAR, 0.05882350",
            "ZAR, CHF, 20.00000000"
    })
    public void testRate(String from, String to, BigDecimal expected) {
        assertEquals(expected, bank.rate(from, to));
    }


    @DisplayName("reduce")
    @ParameterizedTest(name = "reduce from = \"{0}\", amount = {1}, to = \"{2}\", expected = {3}")
    @CsvSource({
            "CHF, 2, USD, 1",
            "ZAR, 20, CHF, 1",
            "ZAR, 17, USD, 1",
            "USD, 5, USD, 5"
            })
    public void testReduce(String from, BigDecimal fromAmt, String to, BigDecimal expected) {
        Money result = bank.reduce(MoneyTest.getCurrencyFactory(from).apply(fromAmt), to);
        assertEquals(MoneyTest.getCurrencyFactory(to).apply(expected), result);
    }

    @DisplayName("reduce with division by zero exception")
    @ParameterizedTest(name = "reduce from = \"{0}\", amount = {1}, to = \"{2}\"")
    @CsvSource({
            "CHF, 1, GBP, 17",
            "ZAR, 1, GBP, 20",
            "USD, 1, INR, 0"
    })
    public void testReduceWithDbzException(String from, BigDecimal fromAmt, String to, int expected) {
        ArithmeticException exception = assertThrows(ArithmeticException.class, () -> {
            bank.reduce(MoneyTest.getCurrencyFactory(from).apply(fromAmt), to);
        });
        assertEquals("Exchange rate not available", exception.getMessage());
    }

    // This error occurs when the currency does not have a factory method.
    @DisplayName("reduce with null pointer exception")
    @ParameterizedTest(name = "reduce from = \"{0}\", amount = {1}, to = \"{2}\"")
    @CsvSource({
            "GBP, 1, ZAR, 20",
            "INR, 1, ZAR, 20",
            "FRF, 1, CHF, 0"
    })
    public void testReduceNotInRateTable(String from, BigDecimal fromAmt, String to, int expected) {
        NullPointerException exception = assertThrows(NullPointerException.class, () -> {
            bank.reduce(MoneyTest.getCurrencyFactory(from).apply(fromAmt), to);
        });
    }


}
